### ISSUE TITLE
RSA Sign fails with only n, e and d component after update from 1.0.2q to 1.1.1d

### ISSUE THREAD

**Issue by seuntjens** [openssl/openssl#10634]
Hi

We've updated from Openssl 1.0.2p to 1.1.1d.
During some testing we've noticed that RSA_sign(...) with private key fails with an unreported error (null). The private RSA key file we use to sign does only contain n, e and d components.
If we 'recover' other components of the RSA key this function does work correctly.

Is it mandatory for RSA keys to contain all compenents since Openssl 1.1.1?
Is there a workaround by configuring to use the SFM algorithm instead of CRT for RSA Encryption?

Kind regards
Jonas

**Comment by mattcaswell** [comment]
> Is it mandatory for RSA keys to contain all compenents since Openssl 1.1.1?

No - this should be ok. Do you have a simple reproducer?

**Comment by seuntjens** [comment]
Oke, so after some rigorous testing, I found out the following.

- Generating an RSA key and copying n,e,d into a new RSA key. Use this new RSA key to sign a message --> goes well! 
- Storing this new (incomplete) RSA key into PKCS1 der file (i2d_RSAPrivateKey_fp) fails with error d2i: field missing

Now what comes next initiates the error:
- We have our own PKCS1 implementation which lets us write a RSA Private key with only n, e, d. If we read this file with d2i_RSAPrivateKey_fp function, no error/fault is given.
- If we try to sign a message using this RSA key --> signing fails.
- However, when we copy the separate components from this RSA key to a new one, with following line, signing goes wel!
 ```
RSA_set0_key( pRsaKeyMissingReadAndCopy, BN_dup(RSA_get0_n(pRsaKeyMissingExtrasRead)), BN_dup(RSA_get0_e(pRsaKeyMissingExtrasRead)), BN_dup(RSA_get0_d(pRsaKeyMissingExtrasRead)) );
 ```

So it seems that the key itself is not entirely corrupted with our own storage function. However when reading this file, the RSA key which comes out of it seems not to be corrupted partly. This was never a problem when we were on OpenSSL 1.0.2p. 

In the attachment you can find a PKCS1 buffer containing a key. Below  the code snippet to reproduce can be found.

 ```
FILE *pTest = nullptr;
pTest = fopen( szTestFile , "r");
RSA* pRsaKeyMissingExtrasRead = RSA_new();
d2i_RSAPrivateKey_fp(pTest, &pRsaKeyMissingExtrasRead);
fclose( pTest );
int ret = RSA_check_key(pRsaKeyMissingExtrasRead);

char message[] = "test data";
unsigned char signature[1024];
unsigned int signatureLen = 0;

if(1 != RSA_sign(NID_sha1, (unsigned char*)message2, strlen (message2), signature2, &signatureLen2, pRsaKeyMissingExtrasRead)) {       
// throw CException("RSA_sign error: %u %s", ERR_get_error(), ERR_error_string(ERR_get_error(), nullptr));
}
ret = RSA_check_key(pRsaKeyMissingExtrasRead);
//TRACE_NEWS("RSA_scheck error after read and get 0 factors: %u %s\n", ERR_get_error(), ERR_error_string(ERR_get_error(), nullptr));

RSA* pRsaKeyMissingReadAndCopy = RSA_new();
RSA_set0_key( pRsaKeyMissingReadAndCopy, BN_dup(RSA_get0_n(pRsaKeyMissingExtrasRead)), BN_dup(RSA_get0_e(pRsaKeyMissingExtrasRead)), BN_dup(RSA_get0_d(pRsaKeyMissingExtrasRead)) );
ret = RSA_check_key(pRsaKeyMissingReadAndCopy);
//TRACE_NEWS("RSA_scheck error after copy of data: %u %s\n", ERR_get_error(), ERR_error_string(ERR_get_error(), nullptr));

if(1 != RSA_sign(NID_sha1, (unsigned char*)message2, strlen (message2), signature2, &signatureLen2, pRsaKeyMissingReadAndCopy)) {
// throw CException("RSA_sign error after copy of data: %u %s", ERR_get_error(), ERR_error_string(ERR_get_error(), nullptr));
}

**Comment by mattcaswell** [comment]
> We have our own PKCS1 implementation which lets us write a RSA Private key with only n, e, d. If we read this file with d2i_RSAPrivateKey_fp function, no error/fault is given.

So, I looked at the key file that you provided and it looks like this:

```` console
$ openssl rsa -in testKey.txt -inform DER -noout -text
Private-Key: (2048 bit)
modulus:
    00:a2:06:cd:85:50:ec:e0:7f:a8:2d:90:f5:3e:b8:
    a0:e1:0e:e9:2b:5a:6d:bc:58:e5:bd:49:22:9f:18:
    33:d0:d0:6b:cf:54:ab:f8:de:92:3d:c7:5e:3b:e5:
    d0:7e:33:e4:9a:93:a9:59:a3:85:c8:48:f9:4f:b0:
    53:0f:03:93:0c:8a:17:bd:19:cb:51:84:38:eb:c3:
    a1:72:00:dd:6e:93:91:ed:5c:96:17:03:7d:28:74:
    2c:c0:64:79:97:dc:5f:3b:b4:7e:3e:81:f6:da:b4:
    e7:75:a6:c4:09:1f:b3:20:9c:9f:4e:06:31:ef:65:
    dc:e6:fd:36:72:28:00:48:d4:91:b0:39:ff:eb:30:
    bd:11:86:3c:78:ce:47:cc:5d:13:06:1f:a8:db:b9:
    10:af:28:f3:89:21:4b:97:9d:1e:2d:e0:a2:39:71:
    a3:2b:87:a6:bc:36:41:6c:a6:99:a1:ef:10:51:0e:
    f1:a4:59:02:30:8d:fb:84:4a:ec:9b:70:6a:a5:19:
    95:32:68:f6:e9:b0:64:49:5d:cf:db:86:81:75:62:
    c8:9d:d8:fe:7e:40:82:af:55:2c:b6:8f:73:2e:ab:
    b3:5c:de:97:61:d6:69:e2:97:5f:87:07:21:10:3d:
    ea:28:cb:9b:f2:3d:6e:6a:6e:d1:fa:b6:8d:b1:96:
    61:ed
publicExponent: 65537 (0x10001)
privateExponent:
    44:1b:4f:f0:a6:1f:ac:0e:7d:07:0a:53:cc:55:97:
    6a:82:f6:08:3f:2e:b9:69:34:c8:e8:21:21:76:ee:
    f0:eb:6b:25:ef:95:8a:3b:24:85:cb:74:15:01:79:
    2e:53:69:d8:80:11:43:98:b4:f7:87:95:24:a6:1d:
    83:c6:20:dd:b8:7e:5a:02:a9:b0:4e:40:0e:fa:60:
    b6:a7:1e:6d:08:a2:e2:7f:74:a7:9f:d3:d0:59:65:
    26:4a:15:c2:7b:e3:9c:19:36:a7:d7:c1:88:15:5a:
    e5:dd:7f:b1:c6:1d:69:ab:ec:5a:c3:c7:e1:b5:93:
    b1:94:e6:79:fd:ce:80:ce:0e:63:00:c5:1b:4b:28:
    4c:01:be:b6:3a:bd:96:6d:fb:59:56:87:4c:65:d4:
    7f:5e:bc:42:6d:ae:5f:09:7b:0e:72:0b:e4:34:90:
    c9:0a:55:60:b0:02:7f:9b:63:26:ae:48:1e:d0:99:
    05:46:6c:6b:30:e3:f4:de:e0:c0:6a:f7:86:7b:08:
    ab:67:54:a2:c2:53:b8:ea:1d:8c:81:66:58:9d:38:
    16:2b:ac:d7:30:8e:44:5d:a4:34:51:43:71:65:a1:
    c7:91:a1:47:4a:b4:bc:42:2d:d0:ce:37:f9:07:33:
    2b:56:04:5d:fa:0b:61:be:c7:3f:97:02:2d:82:c7:
    81
prime1: 0
prime2: 0
exponent1: 0
exponent2: 0
coefficient: 0
````

Note that prime1/prime2 etc are  *present* but set to 0. In fact in the PKCS1 file format these fields are mandatory - so they *must* be present. So, it seems your application is "cheating" by saving the file but setting some mandatory parameters to 0. Therefore when this key is read in by OpenSSL it creates the various BIGNUMs for these fields and sets them to 0. In other words what results is a malformed key.

In both 1.0.2 and 1.1.1 RSA_check_key() fails for these keys because it complains that the "primes" are in fact not prime (which is true, because they are 0).

There have been some optimisations in the codebase for RSA signing between 1.0.2 and 1.1.1. Previously the signing step didn't need to use p/q etc. It still doesn't in 1.1.1 - but if they are present then it will use them. In the case of a key loaded from your file it checks to see if the values are there; sees that they are and then uses them. But because they are incorrect it fails. In the case of a new key that you construct which only has the n, e and d parameters signing works correctly because it checks to see if p/q are present, sees that they aren't and then falls back to the old way of doing things.

So, in other words, I conclude that OpenSSL is doing the right thing here. The problem is that you have a malformed key file. It so happens that in 1.0.2 you were "lucky" and got away with it, but in 1.1.1 this is no longer the case.

**Comment by seuntjens** [comment]
> So, in other words, I conclude that OpenSSL is doing the right thing here. The problem is that you have a malformed key file. It so happens that in 1.0.2 you were "lucky" and got away with it, but in 1.1.1 this is no longer the case.

I agree. The format we use does not seem to match with the mandatory PKCS1 Fields.

Thanks for the clarification!

**Comment by levitte** [comment]
It's a bit interesting that the ASN.1 structure *requires* the CRT numbers to be present, while the cryptographic primitives do not (they will take advantage of them *if* they are present, though): https://tools.ietf.org/html/rfc8017#section-5. I can see how that confuses things...
